bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java

```diff
@@ -0,0 +1,356 @@
+
+        startAutomaticRefresh();
+        logger.debug("Finished initializing!");
+    }
+
+    private String getUrl(String type) {
+        String urlStr = type;
+        return urlStr.replace("%IP%", StringUtils.trimToEmpty(config.ip));
+    }
+
+    private void sendData(String key, String value) {
+        String errorMsg = null;
+        String urlStr = getUrl(GoEChargerBindingConstants.MQTT_URL).replace("%KEY%", key).replace("%VALUE%", value);
+        logger.debug("POST URL = {}", urlStr);
+
+        String result = null;
+        try {
+            ContentResponse contentResponse = httpClient.newRequest(urlStr).method(HttpMethod.POST)
+                    .timeout(5, TimeUnit.SECONDS).send();
+
+            result = contentResponse.getContentAsString();
+            // TODO check if value really changed
+            logger.debug("Response: {}", result);
+        } catch (Exception e) {
+            errorMsg = e.getMessage();
+        }
+
+        if (result != null) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+            return;
+        }
+
+        logger.warn("Error in Go-eCharger request: {}", errorMsg);
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, errorMsg);
+    }
+
+    /**
+     * Request new data from Go-E charger
+     *
+     * @return the Go-E charger object mapping the JSON response or null in case of
+     *         error
+     */
+    @Nullable
+    private GoEStatusResponse getGoEData() {
+        GoEStatusResponse result = null;
+        String errorMsg = null;
+
+        String urlStr = getUrl(GoEChargerBindingConstants.API_URL);
+        logger.debug("GET URL = {}", urlStr);
+
+        try {
+            ContentResponse contentResponse = httpClient.newRequest(urlStr).method(HttpMethod.GET)
+                    .timeout(5, TimeUnit.SECONDS).send();
+
+            result = gson.fromJson(contentResponse.getContentAsString(), GoEStatusResponse.class);
+            // logger.debug("Response: {}", contentResponse.getContentAsString());
+        } catch (Exception e) {
+            errorMsg = e.getMessage();
+        }
+
+        if (result != null) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+            return result;
+        }
+
+        retryCounter++;
+        if (retryCounter == 1) {
+            logger.warn("Error in getting data from Go-E charger, retrying once");
+            return getGoEData();
+        }
+        logger.warn("Error in Go-E charger response: {}", errorMsg);
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, errorMsg);
+        return null;
+    }
+
+    private void refresh() {
+        // Request new GoE data
+        retryCounter = 0;
+        goeResponse = getGoEData();
+
+        // Update all channels from the updated GoE data
+        getThing().getChannels().forEach(channel -> updateChannel(channel.getUID().getId()));
```

Comment:
```suggestion
        getThing().getChannels().map(channel -> channel.getUID().getId()).forEach(id -> updateState(id, getValue(id)));
```

You can the remove the updateState method. If you check null on `goeReponse` (and return if empty) above and add it as parameter to `getValue` (making it `private State getValue(String id, GoEStatusResponse goeResponse)`, you can also remove the field, making the whole thing a bit more thread-safe.

Reply:
If i check for null on goeResponse i also need to set null on all channels, which updateChannel already does for me

Reply:
goeResponse is also used in RefreshCommand, can i ignore the refresh command? then this could work.